plugins/modules/source_control/gitlab/gitlab_project_variable.py

```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
+    # but it is required  and only relevant for check mode!!
+    # logic represents state 'present' when not purge. all other can be derived from that
+    # untouched => equal in both
+    # updated => name and scope are equal
+    # added => name and scope does not exist
+    untouched = list()
+    updated = list()
+    added = list()
+
+    if state == 'present':
+        existing_key_scope_vars = list()
+        for item in existing_variables:
+            existing_key_scope_vars.append({'key': item.get('key'), 'environment_scope': item.get('environment_scope')})
+
+        for idx in range(len(requested_variables)):
+            if requested_variables[idx] in existing_variables:
+                untouched.append(requested_variables[idx])
+            else:
+                compare_item = {'key': requested_variables[idx].get('name'), 'environment_scope': requested_variables[idx].get('environment_scope', '*')}
+                if compare_item in existing_key_scope_vars:
+                    updated.append(requested_variables[idx])
+                else:
+                    added.append(requested_variables[idx])
+
+    return untouched, updated, added
 
 
-def native_python_main(this_gitlab, purge, var_list, state, module):
+def native_python_main(this_gitlab, purge, requested_variables, state, module):
 
     change = False
     return_value = dict(added=list(), updated=list(), removed=list(), untouched=list())
 
     gitlab_keys = this_gitlab.list_all_project_variables()
-    existing_variables = [x.get_id() for x in gitlab_keys]
-
-    for key in var_list:
-
-        if isinstance(var_list[key], string_types) or isinstance(var_list[key], (integer_types, float)):
-            value = var_list[key]
-            masked = False
-            protected = False
-            variable_type = 'env_var'
-            environment_scope = None
-        elif isinstance(var_list[key], dict):
-            value = var_list[key].get('value')
-            masked = var_list[key].get('masked', False)
-            protected = var_list[key].get('protected', False)
-            variable_type = var_list[key].get('variable_type', 'env_var')
-            environment_scope = var_list[key].get('environment_scope')
+    before = [x.attributes for x in gitlab_keys]
+
+    gitlab_keys = this_gitlab.list_all_project_variables()
+    existing_variables = [x.attributes for x in gitlab_keys]
+
+    # preprocessing:filter out and enrich before compare
+    for item in existing_variables:
+        item.pop('project_id')
+
+    for item in requested_variables:
+        item['key'] = item.pop('name')
+        item['value'] = str(item.get('value'))
+        if item.get('protected') is None:
+            item['protected'] = False
+            module.deprecate("The 'protected' attribute will be set to 'True' if not defined.",
+                             version='5.0.0', collection_name='community.general')
```

Comment:
```suggestion
                             version='6.0.0', collection_name='community.general')
```
The changelog fragment needs a `deprecated_features` part then.

I don't really get why protected should be set to `true` if not defined? What is the rationale behind that?

Also, why does this have to be done in this PR, which already changes quite a lot? Is this really necessary?

Reply:
> I don't really get why protected should be set to true if not defined? What is the rationale behind that?

It's the Web GUI default behaviour and kind of a security feature (_mitigation of accidental credentials leak on unprotected branches, such as merge requests_)

> Also, why does this have to be done in this PR, which already changes quite a lot? Is this really necessary?

Sure, we can change this in a later PR.